cpp/src/fil/infer.cu

```diff
@@ -248,20 +248,33 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_FEW_CLASSES> : finalize_block {
     __syncthreads();  // free up input row
     // load margin into shared memory
     auto per_thread = (vec<NITEMS, float>*)tmp_storage;
-    if (threadIdx.x >= num_classes) per_thread[threadIdx.x] = acc;
+    per_thread[threadIdx.x] = acc;
 
     __syncthreads();
     // reduce per-thread margin summand into per-class complete margin
     // (for each of the NITEMS rows)
-    // TODO(levsnv): use CUB/tree reduction when num_classes is small
-    if (threadIdx.x < num_classes) {
-      for (int c = threadIdx.x + num_classes; c < blockDim.x; c += num_classes)
-        acc += per_thread[c];
+    int t = threadIdx.x;
+    int sets = blockDim.x / num_classes;
```

Comment:
I think it's better to have more documentation comments, to have the reduction code easier to understand.